### PR TITLE
DI – Runtime: ServiceContainer, ServiceResolver, and Bootstrap Helpers

### LINKED ISSUE #885: DI - Runtime: ServiceContainer, ServiceResolver, and Bootstrap Helpers
# Technical Requirements Document: DI – Runtime: ServiceContainer, ServiceResolver, and Bootstrap Helpers

**Project:** Tiferet Framework  
**Repository:** https://github.com/greatstrength/tiferet  
**Date:** June 30, 2026  
**Version:** Core DDD Parity II – Events and Repos (tracking milestone)

## 1. Overview

The `tiferet/di` package is the framework's dependency-injection runtime. Today it exposes only the `ServiceProvider` abstract base and its `DynamicServiceProvider` implementation, which is insufficient for the events- and bootstrap-layer parity targeted by this milestone. This story adds the DI runtime engine and application service provider — `ServiceContainer` and `ServiceResolver` — together with four module-level bootstrap helpers (`injectable_parameter_names`, `normalize_flags`, `create_cache_key`, `merge_settings`) to `tiferet/di/settings.py`, and exports all six from the package.

`ServiceContainer` is the low-level engine that registers service types and constants and instantiates services with their constructor parameters wired to sibling registrations, backed by the `dependency_injector` `DynamicContainer`. `ServiceResolver` is the application service provider: it takes a `DIService` directly (in the spirit of a domain event), accepts an injected parameter parser so the DI layer never imports the events package, merges typed bootstrap defaults beneath the repository's settings, and builds and caches a per-flag `ServiceContainer`.

The change is **purely additive**: `ServiceProvider`/`DynamicServiceProvider` remain in place so existing bootstrap and context consumers continue to function. Their eventual removal, and the rewiring of the contexts/blueprints/assets layers onto `ServiceResolver`, are deferred companion work outside this story. This runtime is the hard prerequisite that unblocks #863 (`CreateServiceResolver`), which imports `ServiceResolver` and `injectable_parameter_names` from `tiferet.di`.

### In Scope
- **Add** four module-level functions to `tiferet/di/settings.py`: `injectable_parameter_names`, `normalize_flags`, `create_cache_key`, `merge_settings`.
- **Add** the `ServiceContainer` class (low-level DI engine backed by `dependency_injector`'s `DynamicContainer`).
- **Add** the `ServiceResolver` class (application service provider taking a `DIService`, an injected `parse_parameter`, and typed bootstrap defaults).
- **Update** `tiferet/di/__init__.py` to export the six new artifacts while retaining the existing exports.
- **Add** the `tests/di/` package (`__init__.py` + `test_settings.py`) covering the new runtime.

### Out of Scope
- The `CreateServiceResolver` bootstrap event (#863) — it consumes this runtime; separate story.
- **Removal** of `ServiceProvider` / `DynamicServiceProvider`, and deduplicating `DynamicServiceProvider.build_factory` onto the shared `injectable_parameter_names` helper.
- Rewiring runtime consumers onto `ServiceResolver`/`ServiceContainer`: `tiferet/contexts/di.py` (`DIContext`), `tiferet/contexts/feature.py`, `tiferet/blueprints/main.py`, `tiferet/blueprints/cli.py`, `tiferet/assets/blueprints.py`.
- Resolving the pre-existing retired-name import warning originating in `tiferet/contexts/di.py`.
- Relocating the dormant co-located `tiferet/di/tests/test_dynamic.py` (owned by the Tests relocation story #865).

## 3. Components Affected

| Component | File/Path | Artifact action |
|-----------|-----------|-----------------|
| DI runtime | `tiferet/di/settings.py` | **Add** a `# *** functions` section (4 helpers) and two classes (`ServiceContainer`, `ServiceResolver`) under `# *** classes`, plus the supporting imports; existing `ServiceProvider` is unchanged |
| DI exports | `tiferet/di/__init__.py` | **Update** to export the six new artifacts; retain `ServiceProvider`, `DynamicServiceProvider`, and the `DependenciesServiceProvider` alias |
| DI tests | `tests/di/__init__.py`, `tests/di/test_settings.py` | **Add** the new test package and runtime test suite |

### 4.1 Add — module-level functions in `tiferet/di/settings.py`

Add the supporting imports: `inspect`; `typing` (`Any, Callable, Dict, List, Tuple`); `from dependency_injector import containers, providers`; `from ..domain import ServiceRegistration`; `from ..interfaces.di import DIService`. Organize under a `# *** functions` section:

- `injectable_parameter_names(service_type: type) -> List[str]` — return the injectable constructor parameter names, excluding `self` and variadic (`*args`/`**kwargs`) parameters; treat uninspectable types as no-arg (return `[]`).
- `normalize_flags(*flags) -> List[str]` — flatten a mixed sequence of strings, lists, and tuples into a flat list of strings.
- `create_cache_key(flags: List[str] = None) -> str` — return `feature_services`, with `_`-joined flags appended when present.
- `merge_settings(configs=None, constants=None, default_config_index=None, default_constants=None) -> Tuple[List[ServiceRegistration], Dict[str, Any]]` — append default-index entries for any service id not already present, and merge default constants **beneath** repository constants (repository values win).

### 4.2 Add — `ServiceContainer` (low-level DI engine)

Under `# *** classes`, add `ServiceContainer` with attribute `container: containers.DynamicContainer`.

- `__init__(self, services: Dict[str, type] = None)` — create the `DynamicContainer`; register initial services if provided.
- `add_service(self, service_id, service_type)` — register a `Factory` provider built by `build_factory`.
- `add_services(self, services)` — **two passes**: register non-type values as `Object` providers first, then class types as `Factory` providers, so scalar parameter values are available when factory kwargs are wired.
- `add_constants(self, constants)` — register each constant as an `Object` provider.
- `get_service(self, service_id) -> Any` — look up the provider and invoke it; a missing or failing provider raises a **raw exception** for the caller (which has event access) to convert into a structured error.
- `remove_service(self, service_id)` — remove the provider if present; no-op otherwise.
- `build_factory(self, service_type) -> providers.Factory` — wire each injectable parameter (via `injectable_parameter_names`) to a registered sibling provider when one exists.

### 4.3 Add — `ServiceResolver` (application service provider)

Add `ServiceResolver` with attributes `di_service: DIService`, `parse_parameter: Callable`, `default_config_index: Dict[str, ServiceRegistration]`, `default_di_constants: Dict[str, Any]`, and an internal per-flag container cache.

- `__init__(self, di_service, parse_parameter=None, default_config_index=None, default_di_constants=None)` — `parse_parameter` defaults to an identity function (`lambda v: v`) so the DI layer never imports the parameter-parsing event; default collections coerce to `{}`; initialize the container cache.
- `normalize_flags(*flags)` *(static)* — delegate to the module-level helper.
- `create_cache_key(self, flags=None)` — delegate to the module-level helper.
- `list_all_settings(self) -> Tuple[List[ServiceRegistration], Dict[str, Any]]` — read `di_service.list_all()` and merge bootstrap defaults via `merge_settings`.
- `load_constants(self, configurations=None, constants=None, flags=None) -> Dict[str, Any]` — parse top-level constants with `self.parse_parameter`, then merge per-configuration parameters (flagged parameters when a provided flag matches, else the configuration's default parameters).
- `build_type_map(self, configurations=None, flags=None) -> Dict[str, type]` — resolve each configuration's type via `get_service_type(*flags)`; **skip** configurations that resolve to no type (left unregistered, best-case) so an unresolved service later surfaces as a raw resolution error at a layer with event access.
- `build_container(self, flags=None) -> ServiceContainer` — return the cached container for the flag key if present; otherwise load and merge settings, parse constants, build the type map, construct a `ServiceContainer`, register constants then services, cache, and return it.
- `get_dependency(self, registration_id, *flags) -> Any` — normalize flags, build or retrieve the container, and return the resolved service.

### 4.4 Update — `tiferet/di/__init__.py`

Export `ServiceContainer`, `ServiceResolver`, `injectable_parameter_names`, `merge_settings`, `normalize_flags`, and `create_cache_key` (e.g., via `__all__` plus `from .settings import ...`). Retain the existing `ServiceProvider`, `DynamicServiceProvider`, and the `DependenciesServiceProvider` backward-compatible alias.

### 4.5 Add — `tests/di/test_settings.py`

Add a new `tests/di/` package (`__init__.py`) and a standalone pytest module that imports the runtime from `tiferet.di.settings`. Under a `# *** classes` section, add in-test service doubles (a dependency-free service, a constructor-dependency service, and a scalar-parameter service). Cover:

- **`ServiceContainer`**: empty and with-services initialization; `add_service` (including new-instance-per-call); `add_services`; `add_constants` (including injection into a dependent service); `get_service` not-found (raw error); `remove_service` (including nonexistent no-op); cascading dependency injection.
- **`ServiceResolver`**: `create_cache_key`; `build_container` (default, flagged, and cached); skip of a no-type configuration; `get_dependency` (default, varargs, and list-flag forms); `normalize_flags`; `load_constants` (default vs flagged); `list_all_settings` merging defaults; `parse_parameter` injection vs identity default.
- **`merge_settings`**: appends default-only configurations and merges constants with repository priority.

The service doubles referenced by `ServiceRegistration.module_path` must use the importable path `tests.di.test_settings`, so the module must live at `tests/di/test_settings.py` (collected via `testpaths = ["tests", "tests_int"]`).

## 5. Acceptance Criteria

1. `tiferet/di/settings.py` defines `injectable_parameter_names`, `normalize_flags`, `create_cache_key`, `merge_settings`, `ServiceContainer`, and `ServiceResolver`, and the module imports cleanly.
2. `from tiferet.di import ServiceContainer, ServiceResolver, injectable_parameter_names, merge_settings, normalize_flags, create_cache_key` succeeds, and `ServiceProvider`, `DynamicServiceProvider`, and `DependenciesServiceProvider` remain importable from `tiferet.di`.
3. `ServiceResolver` accepts `di_service`, an optional `parse_parameter` (defaulting to identity), `default_config_index`, and `default_di_constants`; `get_dependency(registration_id, *flags)` resolves a configured service, honoring flagged overrides and merged bootstrap defaults.
4. `ServiceContainer.get_service` returns a new instance per call for `Factory`-registered services, returns constants as-is, and raises a raw error for a missing provider (no DI-specific structured error in the engine).
5. `tests/di/test_settings.py` exists, covers the cases in §4.5, and passes under `pytest tests/`, with no regressions in the full suite.
6. The change is additive: `tiferet/contexts/`, `tiferet/blueprints/`, and `tiferet/assets/` are unchanged, and no events-package import is introduced into `tiferet/di`.

## 6. Non-Functional Requirements

- **Consistency:** all artifacts follow the structured code style (artifact comments `# *** imports` / `# ** core` / `# ** infra` / `# ** app` / `# *** functions` / `# *** classes` / `# ** class:` / `# * method:`, RST docstrings, and spacing).
- **Layering:** `tiferet/di` depends on `domain` and `interfaces` only and must not import the events package; the injected `parse_parameter` preserves this boundary.
- **Backward compatibility:** existing `ServiceProvider` / `DynamicServiceProvider` consumers (bootstrap and contexts) continue to function unchanged.
- **No new third-party runtime dependency** is introduced (`dependency-injector` is already a project dependency).

## 7. Prerequisites

| Dependency | Description | Status in `main` |
|------------|-------------|------------------|
| `ServiceRegistration` domain model | `get_dependency(*flags)`, `get_service_type(*flags)`, `parameters`, `id` consumed by the resolver | Present (`tiferet/domain/di.py`) |
| `DIService` contract | `list_all()` returning `(registrations, constants)` | Present (`tiferet/interfaces/di.py`) |
| `dependency_injector` | `DynamicContainer`, `Factory`, and `Object` providers | Present (existing project dependency) |

This TRD is itself the prerequisite that unblocks the Blueprint Events story #863 (`CreateServiceResolver`), which imports `ServiceResolver` and `injectable_parameter_names` from `tiferet.di`.

## 8. Related Code Style Documentation

- [code_style.md](https://github.com/greatstrength/tiferet/blob/main/docs/core/code_style.md) — General structured code style

### PR DESCRIPTION
**Pull request by Ashatz**
## Summary

Adds the additive **DI runtime engine** and **application service provider** to `tiferet/di`, the hard prerequisite that unblocks #863 (`CreateServiceResolver`). Implements the TRD in #885.

Part of milestone #29 (Core DDD Parity II – Events and Repos). Addresses #885.

## What changed

`tiferet/di/settings.py` (additive — `ServiceProvider` unchanged):

- **`# *** functions`** — `injectable_parameter_names` (constructor params minus `self`/variadics; uninspectable → `[]`), `normalize_flags` (flatten strings/lists/tuples), `create_cache_key` (`feature_services[_<flags>]`), and `merge_settings` (append default-index entries for absent ids; merge default constants **beneath** repository constants).
- **`ServiceContainer`** — low-level engine backed by `dependency_injector`'s `DynamicContainer`: two-pass `add_services` (scalars as `Object` before types as `Factory`), `add_constants`, `get_service` that raises a **raw** error for a missing/failing provider (no structured `TiferetError` in the engine), idempotent `remove_service`, and `build_factory` wired via `injectable_parameter_names`.
- **`ServiceResolver`** — application service provider taking a `DIService` directly plus an injected `parse_parameter` (identity default, so `tiferet/di` never imports the events package) and typed bootstrap defaults. Provides `normalize_flags` (static), `create_cache_key`, `list_all_settings`, `load_constants`, `build_type_map` (skips no-type registrations), `build_container` (per-flag cache), and `get_dependency(registration_id, *flags)`.

`tiferet/di/__init__.py` — exports the six new artifacts via `__all__` while retaining `ServiceProvider`, `DynamicServiceProvider`, and the `DependenciesServiceProvider` alias.

`tests/di/__init__.py`, `tests/di/test_settings.py` — new standalone suite (34 tests). Doubles use the importable path `tests.di.test_settings` so `ServiceRegistration.module_path` resolves under `testpaths`.

## Acceptance criteria

- [x] AC1 — `settings.py` defines the four helpers + `ServiceContainer` + `ServiceResolver`; module imports cleanly.
- [x] AC2 — the six artifacts import from `tiferet.di`; `ServiceProvider`/`DynamicServiceProvider`/`DependenciesServiceProvider` remain importable.
- [x] AC3 — `ServiceResolver` accepts `di_service`, optional identity `parse_parameter`, `default_config_index`, `default_di_constants`; `get_dependency` honors flagged overrides and merged defaults.
- [x] AC4 — `ServiceContainer.get_service` returns a new instance per `Factory` call, returns constants as-is, and raises a raw error for a missing provider.
- [x] AC5 — `tests/di/test_settings.py` covers §4.5 and passes under `pytest tests/` with no regressions.
- [x] AC6 — additive only: `tiferet/contexts/`, `tiferet/blueprints/`, `tiferet/assets/` unchanged; no events-package import added to `tiferet/di`.

## Testing

- `pytest tests/di` → **34 passed**
- `pytest tests/` → **742 passed, 11 skipped** (708 baseline + 34 new; no regressions)

## Notes

- Purely additive; rewiring contexts/blueprints/assets onto `ServiceResolver`/`ServiceContainer` and removing `ServiceProvider`/`DynamicServiceProvider` are deferred companion work (out of scope per the TRD).
- The pre-existing import-time warning from `tiferet/contexts/di.py` (`ServiceConfiguration`) is unchanged by this PR — it originates from deferred contexts rewiring, not `tiferet/di`.

[Warp conversation](https://app.warp.dev/conversation/7b30fb3e-5ba0-4dc9-8743-7e2c4939fcd9)

Co-Authored-By: Oz <oz-agent@warp.dev>

### CODE REVIEW

File: tiferet/di/settings.py [hunk 1]
```diff
@@ -2,8 +2,139 @@
...
+    normalized = []
+    for flag in flags:
+
+        # Expand nested list/tuple flag groups.
+        if isinstance(flag, (list, tuple)):
+            normalized.extend(flag)
+
+        # Append non-empty scalar flags.
+        elif flag is not None:
+            normalized.append(flag)
```
Comment (Ashatz) [inline tiferet/di/settings.py]: `normalize_flags` diverges from the canonical `v2.0-proto` implementation in two ways:

1. **No `str()` coercion.** The canonical version stringifies every flag (`str(f)` in the list/tuple branch, `str(flag)` for scalars). Here flags are appended as-is, so the documented `-> List[str]` return contract is violated for non-string flags (e.g. ints/enums).
2. **Inconsistent `None` handling.** `elif flag is not None` drops only *scalar* `None`; a `None` nested inside a list/tuple is still passed through by `extend(flag)`.

Downstream impact: `create_cache_key` does `'_'.join(flags)`, which raises `TypeError` for a non-string flag or a `None` nested in a list. Canonical avoids this by stringifying. Suggested:

```python
for flag in flags:
    if isinstance(flag, (list, tuple)):
        normalized.extend(str(f) for f in flag)
    else:
        normalized.append(str(flag))
```

If the new behavior is intentional, please add a test for non-string/`None` flags (currently uncovered in both suites).

**Review by Ashatz — COMMENTED** [review]
## Review — issue #885 (evaluated against `v2.0-proto` as source of truth)

Scope limited to the four issue-885 files. Tests pass (34) and AC1–AC6 hold. The additive retention of `ServiceProvider` / `DynamicServiceProvider` / `DependenciesServiceProvider` (dropped in `v2.0-proto`) is intentional for the `main` target and is **not** flagged.

### Intended fixes

**Inline (see comments):**
1. `normalize_flags` — restore `str()` coercion and consistent `None` handling to match canonical; honors the `List[str]` contract and avoids a `TypeError` in `create_cache_key`. Add a non-string/`None` flag test if the behavior is kept.
2. Cache attribute `container_cache` → align with canonical private `_containers` (or confirm the public name is intentional).

**Nits (optional — fidelity to `v2.0-proto` / conventions):**
3. Add module docstring `"""Tiferet DI Settings"""` at the top of `tiferet/di/settings.py`.
4. `build_container` — guard with `if type_map:` before `add_services(type_map)` (canonical; behavior is equivalent since an empty map is a no-op).
5. Resolver default coercion — use `... is not None else {}` for `default_config_index` / `default_di_constants` (canonical idiom).
6. Tests — canonical and AGENTS.md convention use `mock.Mock(spec=DIService)`; the hand-rolled `FakeDIService` adds ~100 lines of boilerplate. Optional.

[Warp conversation](https://app.warp.dev/conversation/99cf975a-83f0-478a-8204-0cae30b24acf)

File: tiferet/di/settings.py [hunk 2]
```diff
@@ -76,3 +207,387 @@ def remove_service(self, service_id: str):
...
+    parse_parameter: Callable
+
+    # * attribute: default_config_index
+    default_config_index: Dict[str, ServiceRegistration]
+
+    # * attribute: default_di_constants
+    default_di_constants: Dict[str, Any]
+
+    # * attribute: container_cache
+    container_cache: Dict[str, ServiceContainer]
```
Comment (Ashatz) [inline tiferet/di/settings.py]: Naming divergence from the canonical `v2.0-proto` resolver, which stores the per-flag cache as a private `self._containers` (initialized in `__init__`, with no class-level attribute). Here it is a public, declared `container_cache` (also referenced in `__init__` and `build_container`). Functionally fine and internally consistent, but the deferred contexts/blueprints rewiring may expect `_containers` — aligning now avoids churn in the companion PR. If the public name is intentional, feel free to ignore.
